rfcs/019 - Datastore Compute RFC.md

```diff
@@ -0,0 +1,202 @@
+# 019 - Datastore Compute RFC
+
+## Current Status
+
+### Proposed
+
+2021-07-01
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Luis De Pombo <luis@alantechnologies.com>
+- Alejandro Guillen <alejandro@alantechnologies.com>
+- Colton Donnelly <colton@alantechnologies.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/alantech/alan/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- David Ellis <david@alantechnologies.com>
+
+## Summary
+
+Often we want to extract small amounts of data out of a larger pool of data. Pushing that work to the data rather than transferring the data to where the answer is desired and then computing it is more efficient from both a time and bandwidth perspective in many situations. This is also a parallelization strategy if the relevant data is spread across multiple nodes, and complements the event-level and array-level parallelization already present in Alan.
+
+Adding the right remote compute primitive on top of datastore will allow us to build lots of data primitives on top efficiently. We should be able to build queues, B-Tree indexes, tables, MapReduce, etc with code that *looks like* single-threaded, single-process algorithmic tutorial code that is actually multi-server. Entire classes of external tooling simply fall away when you can express what you need from them cleanly in a few lines of code and don't have to deploy anything extra to support them! :)
+
+## Expected SemBDD Impact
+
+This would be a minor update if we were post-1.0 as it should have zero breaking impact on existing code.
+
+## Proposal
+
+When Alan has user-defined code passed into a built-in opcode, such as the `Array<T>`'s `map`, the inner function is compiled into AGA as a `closure for <uuid>` that temporarily takes over the outer scope's `HandlerMemory` and executes the inner function with the relevant arguments passed in on each run, then the output is placed into the `HandlerMemory` and it continues on. If `mapLin` was used, it runs sequentially and the inner function can mutate values in the outer scope safely (and is allowed to do so).
+
+There is no theoretical reason why `map` must do its work on the same machine in the cluster. It could copy the `HandlerMemory` to any other node, do its work there, and then copy it back at the end. This simply doesn't make any sense from a performance perspective since you must spend the time to serialize, transmit, deserialize, compute, serialize, transmit, and finally deserialize, while staying on the same machine only requires creating a special child `HanlderMemory` associated with the original one, almost no overhead versus the compute step.
```

Comment:
typo: `HanlderMemory`

Reply:
```suggestion
There is no theoretical reason why `map` must do its work on the same machine in the cluster. It could copy the `HandlerMemory` to any other node, do its work there, and then copy it back at the end. This simply doesn't make any sense from a performance perspective since you must spend the time to serialize, transmit, deserialize, compute, serialize, transmit, and finally deserialize, while staying on the same machine only requires creating a special child `HandlerMemory` associated with the original one, almost no overhead versus the compute step.
```